cpp/community_detection_module/algorithm/label_propagation.cpp

```diff
@@ -0,0 +1,291 @@
+}
+
+bool LabelRankT::distinct_enough(mg_graph::Node<uint64_t> node_i,
+                                 double similarity_threshold) {
+  std::vector<std::uint64_t> labels_i = most_probable_labels(node_i.id);
+  std::uint64_t label_similarity = 0;
+
+  int node_i_degree = 0;
+  for (const auto node_j : graph->Neighbours(node_i.id)) {
+    node_i_degree++;
+
+    std::vector<std::uint64_t> labels_j = most_probable_labels(node_j.node_id);
+
+    if (is_subset(labels_i, labels_j)) label_similarity++;
+  }
+
+  if (label_similarity <= node_i_degree * similarity_threshold) return true;
+
+  return false;
+}
+
+std::unordered_map<std::uint64_t, double> LabelRankT::propagate(
+    mg_graph::Node<uint64_t> node_i) {
+  std::unordered_map<std::uint64_t, double> new_label_Ps_i;
+
+  // propagate own probabilities (handle self-loops)
+  for (const auto [label, P] : label_Ps[node_i.id]) {
+    new_label_Ps_i.insert({label, w_selfloop / sum_w[node_i.id] * P});
+  }
+
+  // propagate neighbors’ probabilities
+  for (const auto node_j : graph->Neighbours(node_i.id)) {
+    double contribution = DEFAULT_WEIGHT / sum_w[node_i.id];
+
+    for (const auto [label, P] : label_Ps[node_j.node_id]) {
+      new_label_Ps_i[label] += contribution * P;
+    }
+  }
+
+  return new_label_Ps_i;
+}
+
+void LabelRankT::inflate(
+    std::unordered_map<std::uint64_t, double>& node_label_Ps, double exponent) {
+  double sum_Ps = 0;
+
+  for (const auto [label, _] : node_label_Ps) {
+    double inflated_node_label_Ps = pow(node_label_Ps[label], exponent);
+    node_label_Ps[label] = inflated_node_label_Ps;
+    sum_Ps += inflated_node_label_Ps;
+  }
+
+  for (const auto [label, _] : node_label_Ps) node_label_Ps[label] /= sum_Ps;
+
+  return;
+}
+
+void LabelRankT::cutoff(
+    std::unordered_map<std::uint64_t, double>& node_label_Ps,
+    double min_value) {
+  std::vector<std::uint64_t> to_be_removed;
+
+  for (const auto [label, P] : node_label_Ps) {
+    if (P < min_value) to_be_removed.push_back(label);
+  }
+
+  for (const auto label : to_be_removed) node_label_Ps.erase(label);
+
+  return;
+}
+
+LabelRankT::LabelRankT(std::unique_ptr<mg_graph::Graph<>>& graph,
+                       double w_selfloop, double similarity_threshold,
+                       double exponent, double min_value)
+    : graph(graph) {
+  this->w_selfloop = w_selfloop;
+  this->similarity_threshold = similarity_threshold;
+  this->exponent = exponent;
+  this->min_value = min_value;
+}
+
+std::pair<bool, std::uint64_t> LabelRankT::iteration(
+    bool incremental, std::unordered_set<std::uint64_t> changed_nodes) {
+  std::unordered_map<std::uint64_t, std::unordered_map<std::uint64_t, double>>
+      updated_label_Ps;
+
+  bool none_updated = true;
+  std::uint64_t most_updates = 0;
+
+  for (const auto node : graph->Nodes()) {
+    if (incremental) {
+      bool was_updated = changed_nodes.count(node.id) > 0 ? true : false;
+      if (not was_updated) continue;
+    }
```

Comment:
```suggestion
    bool was_updated = changed_nodes.count(node.id) > 0 ? true : false;
    if (incremental && !was_updated) {
        continue;
    }
```

Reply:
It seems that you are trying to avoid counting nodes if it is not `incremental`, which also works

Reply:
That was indeed what I was going for!